File: src/test/java/io/github/aalsanie/codes/OutcomeCodeTest.java
package io.github.aalsanie.codes;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

class OutcomeCodeTest {
    @Test
    void createsAndParsesValidCodes() {
        OutcomeCode code = OutcomeCode.of("com.example.payments", "PAYMENT_DECLINED");
        assertEquals("com.example.payments", code.getNamespace());
        assertEquals("PAYMENT_DECLINED", code.getName());
        assertEquals("com.example.payments:PAYMENT_DECLINED", code.getValue());
        assertEquals(code.getValue(), code.toString());
        assertEquals(code, OutcomeCode.parse(code.getValue()));
        assertEquals(code.hashCode(), OutcomeCode.parse(code.getValue()).hashCode());
    }

    @Test
    void rejectsInvalidNamespacesAndNames() {
        assertThrows(NullPointerException.class, () -> OutcomeCode.of(null, "CODE"));
        assertThrows(NullPointerException.class, () -> OutcomeCode.of("com.example", null));
        assertThrows(IllegalArgumentException.class, () -> OutcomeCode.of("", "CODE"));
        assertThrows(IllegalArgumentException.class, () -> OutcomeCode.of("a".repeat(129), "CODE"));
        assertThrows(IllegalArgumentException.class, () -> OutcomeCode.of("com..example", "CODE"));
        assertThrows(IllegalArgumentException.class, () -> OutcomeCode.of("a".repeat(64), "CODE"));
        assertThrows(IllegalArgumentException.class, () -> OutcomeCode.of("1example", "CODE"));
        assertThrows(IllegalArgumentException.class, () -> OutcomeCode.of("com.example-", "CODE"));
        assertThrows(IllegalArgumentException.class, () -> OutcomeCode.of("Com.example", "CODE"));
        assertThrows(IllegalArgumentException.class, () -> OutcomeCode.of("com.example_foo", "CODE"));
        assertThrows(IllegalArgumentException.class, () -> OutcomeCode.of("com.example", ""));
        assertThrows(IllegalArgumentException.class, () -> OutcomeCode.of("com.example", "A".repeat(65)));
        assertThrows(IllegalArgumentException.class, () -> OutcomeCode.of("com.example", "bad"));
        assertThrows(IllegalArgumentException.class, () -> OutcomeCode.of("com.example", "BAD-NAME"));
    }

    @Test
    void parseRejectsMalformedValues() {
        assertThrows(NullPointerException.class, () -> OutcomeCode.parse(null));
        assertThrows(IllegalArgumentException.class, () -> OutcomeCode.parse("com.example"));
        assertThrows(IllegalArgumentException.class, () -> OutcomeCode.parse(":CODE"));
        assertThrows(IllegalArgumentException.class, () -> OutcomeCode.parse("com:example:CODE"));
        assertThrows(IllegalArgumentException.class, () -> OutcomeCode.parse("com.example:"));
        assertThrows(
            NullPointerException.class,
            () -> OutcomeCode.parseOrNull(null)
        );
        assertNull(OutcomeCode.parseOrNull("bad"));
    }

    @Test
    void identityAndOrderingUseBothParts() {
        OutcomeCode a = OutcomeCode.of("com.example", "A");
        OutcomeCode b = OutcomeCode.of("com.example", "B");
        OutcomeCode otherNamespace = OutcomeCode.of("org.example", "A");
        assertTrue(a.compareTo(b) < 0);
        assertNotEquals(a, b);
        assertNotEquals(a, otherNamespace);
        assertNotEquals(a, "com.example:A");
        assertThrows(NullPointerException.class, () -> a.compareTo(null));
    }
}

File: src/test/kotlin/io/github/aalsanie/codes/OutcomeCodeTest.kt
package io.github.aalsanie.codes

import org.junit.jupiter.api.Test

class OutcomeCodeTest {
    @Test fun createsValidCode() {
        val code = OutcomeCode.of("com.example.payments", "PAYMENT_DECLINED")
        assertEquals("com.example.payments", code.namespace)
        assertEquals("PAYMENT_DECLINED", code.name)
        assertEquals("com.example.payments:PAYMENT_DECLINED", code.value)
        assertEquals(code.value, code.toString())
    }

    @Test fun parsesValidCode() = assertEquals(
        OutcomeCode.of("com.example", "NOT_FOUND"),
        OutcomeCode.parse("com.example:NOT_FOUND"),
    )

    @Test fun parseOrNullReturnsNullForInvalidCode() {
        assertNull(OutcomeCode.parseOrNull("invalid"))
        assertNull(OutcomeCode.parseOrNull("com.example:bad"))
    }

    @Test fun rejectsMissingDelimiter() { assertFails<IllegalArgumentException> { OutcomeCode.parse("com.example") } }
    @Test fun rejectsMultipleDelimiters() { assertFails<IllegalArgumentException> { OutcomeCode.parse("com:example:CODE") } }
    @Test fun rejectsEmptyNamespace() { assertFails<IllegalArgumentException> { OutcomeCode.of("", "CODE") } }
    @Test fun rejectsEmptyNamespaceSegment() { assertFails<IllegalArgumentException> { OutcomeCode.of("com..example", "CODE") } }
    @Test fun rejectsUppercaseNamespace() { assertFails<IllegalArgumentException> { OutcomeCode.of("Com.example", "CODE") } }
    @Test fun rejectsNamespaceSegmentStartingWithDigit() { assertFails<IllegalArgumentException> { OutcomeCode.of("com.1example", "CODE") } }
    @Test fun rejectsNamespaceSegmentEndingWithHyphen() { assertFails<IllegalArgumentException> { OutcomeCode.of("com.example-", "CODE") } }
    @Test fun rejectsNamespaceIllegalCharacter() { assertFails<IllegalArgumentException> { OutcomeCode.of("com.example_foo", "CODE") } }
    @Test fun rejectsNamespaceTooLong() { assertFails<IllegalArgumentException> { OutcomeCode.of("a".repeat(129), "CODE") } }
    @Test fun rejectsNamespaceSegmentTooLong() { assertFails<IllegalArgumentException> { OutcomeCode.of("a".repeat(64), "CODE") } }
    @Test fun rejectsEmptyName() { assertFails<IllegalArgumentException> { OutcomeCode.of("com.example", "") } }
    @Test fun rejectsLowercaseName() { assertFails<IllegalArgumentException> { OutcomeCode.of("com.example", "bad") } }
    @Test fun rejectsNameIllegalCharacter() { assertFails<IllegalArgumentException> { OutcomeCode.of("com.example", "BAD-NAME") } }
    @Test fun rejectsNameTooLong() { assertFails<IllegalArgumentException> { OutcomeCode.of("com.example", "A".repeat(65)) } }

    @Test fun equalityUsesStructuredIdentity() {
        val first = OutcomeCode.of("com.example", "CODE")
        val second = OutcomeCode.of("com.example", "CODE")
        assertEquals(first, second)
        assertEquals(first.hashCode(), second.hashCode())
    }

    @Test fun inequalityUsesBothParts() {
        assertNotEquals(OutcomeCode.of("com.example", "A"), OutcomeCode.of("com.example", "B"))
        assertNotEquals(OutcomeCode.of("com.example", "A"), OutcomeCode.of("org.example", "A"))
    }

    @Test fun comparisonUsesSerializedValue() {
        val first = OutcomeCode.of("com.example", "A")
        val second = OutcomeCode.of("com.example", "B")
        assertTrue(first < second)
    }
}
